File: frontend/src/utils/declarationWorkflow.ts
// 与后端共用申报专用模板；版本一经使用不可原地改动，应新增版本。
import catalog from '@/config/declaration-workflows.json'

export interface DeclarationNode {
  code: string
  title: string
  roleKeys: string[]
  kind: string
  evidence: boolean
}
export type WorkflowSource = {
  channelCode?: string; channelName?: string; flowNodes?: string
  needApproval?: number; status?: string; flowNode?: string
  posts?: Record<string, string | undefined>
}
export function selectDeclarationWorkflow(source: WorkflowSource = {}): string {
  const byCode = (catalog.channels as Record<string, string>)[source.channelCode || '']
  const byName = (catalog.names as Record<string, string>)[source.channelName || '']
  // 有独立审签模板的渠道不能由客户端 needApproval=0 绕过。
  if (byCode && byCode !== 'common-v1') return byCode
  if (byName) return byName
  if (/无需审批|直接报备/.test(source.flowNodes || '')) return 'report-v1'
  return byCode || 'common-v1'
}
export function declarationWorkflowId(source: WorkflowSource = {}): string {
  const version = source.posts?.__workflow
  if (version) {
    if (!(version in catalog.templates)) throw new Error(`不支持的申报流程版本：${version}`)
    return version
  }
  // 已存在的项目只要带有渠道编码/名称，就按该渠道展示和执行对应的申报链。
  // 只有完全没有渠道信息的历史记录才使用旧固定链兜底。
  if (source.channelCode || source.channelName) return selectDeclarationWorkflow(source)
  if (source.status && !['DRAFT', 'REJECTED'].includes(source.status)) {
    return source.needApproval === 0 ? 'legacy-report-v0' : 'legacy-v0'
  }
  return selectDeclarationWorkflow(source)
}
export function declarationNodes(source: WorkflowSource = {}): DeclarationNode[] {
  const ids = (catalog.templates as Record<string, string[]>)[declarationWorkflowId(source)]
  return ids.map(code => catalog.nodes.find(node => node.code === code)!)
}
export function declarationAuditNodes(source: WorkflowSource = {}) {
  return declarationNodes(source).filter(node => node.kind !== 'fill' && node.kind !== 'end')
}
export function currentDeclarationNode(source: WorkflowSource) {
  const title = source.flowNode === '承办部门负责人' ? '项目承担部门负责人' : source.flowNode
  return declarationAuditNodes(source).find(node => node.title === title)
}
export function declarationActorLabels(source: WorkflowSource): string[] {
  return (currentDeclarationNode(source)?.roleKeys || []).map(key => source.posts?.[key]).filter(Boolean) as string[]
}
export function declarationRequiredPosts(source: WorkflowSource) {
  const seen = new Set<string>()
  return declarationNodes(source).flatMap(node => node.roleKeys.slice(0, 1).flatMap(key => {
    if (seen.has(key)) return []
    seen.add(key)
    return [{ key, label: node.title }]
  }))
}

File: backend/src/main/java/com/comac/rpm/modules/declaration/DeclarationWorkflow.java
package com.comac.rpm.modules.declaration;

import com.comac.rpm.common.BusinessException;
import com.comac.rpm.modules.declaration.entity.ProjDeclaration;
import com.comac.rpm.modules.dict.entity.ProjChannel;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.springframework.core.io.ClassPathResource;
import org.springframework.stereotype.Component;
import java.io.IOException;
import java.util.List;
import java.util.Map;

/** 仅用于申报。前端和 Mock 直接读取同一个版本化资源。 */
@Component
public class DeclarationWorkflow {
    public record Node(String code, String title, List<String> roleKeys, String kind, boolean evidence) {}
    public record Catalog(Map<String, String> channels, Map<String, String> names,
                          Map<String, List<String>> templates, List<Node> nodes) {}
    private final Catalog catalog;

    public DeclarationWorkflow(ObjectMapper mapper) throws IOException {
        try (var stream = new ClassPathResource("declaration-workflows.json").getInputStream()) {
            catalog = mapper.readValue(stream, Catalog.class);
        }
    }

    public String select(ProjChannel channel) {
        if (channel == null) return "common-v1";
        String byCode = catalog.channels().get(channel.getChannelCode());
        String byName = catalog.names().get(channel.getChannelName());
        if (byCode != null && !"common-v1".equals(byCode)) return byCode;
        if (byName != null) return byName;
        String flow = channel.getFlowNodes() == null ? "" : channel.getFlowNodes();
        if (flow.contains("无需审批") || flow.contains("直接报备")) return "report-v1";
        return byCode == null ? "common-v1" : byCode;
    }

    public String version(ProjDeclaration declaration, Map<String, String> posts) {
        String version = posts.get("__workflow");
        if (version != null && !catalog.templates().containsKey(version))
            throw new BusinessException("不支持的申报流程版本：" + version);
        if (version == null && declaration != null && declaration.getChannelName() != null) {
            String mapped = catalog.names().get(declaration.getChannelName());
            if (mapped != null) return mapped;
        }
        return version == null ? (Integer.valueOf(0).equals(declaration.getNeedApproval())
                ? "legacy-report-v0" : "legacy-v0") : version;
    }

    public List<Node> nodes(String version) {
        return catalog.templates().get(version).stream().map(code -> catalog.nodes().stream()
                .filter(node -> code.equals(node.code())).findFirst().orElseThrow()).toList();
    }

    public List<Node> auditNodes(String version) {
        return nodes(version).stream().filter(n -> !"fill".equals(n.kind()) && !"end".equals(n.kind())).toList();
    }

    public Node current(ProjDeclaration declaration, Map<String, String> posts) {
        String title = "承办部门负责人".equals(declaration.getFlowNode())
                ? "项目承担部门负责人" : declaration.getFlowNode();
        return auditNodes(version(declaration, posts)).stream().filter(n -> n.title().equals(title))
                .findFirst().orElse(null);
    }
}
